File: code/scenes/src/main/java/org/fbme/scenes/controllers/SceneViewpoint.java
package org.fbme.scenes.controllers;

import java.awt.*;

public interface SceneViewpoint {

    Point getEditorShift();

    float getEditorScale();

    default int toEditorDimension(int d) {
        return (int) (d * getEditorScale());
    }

    default int fromEditorDimension(int d) {
        return (int) (d / getEditorScale());
    }

    default int translateToEditorX(int x) {
        return toEditorDimension(x) + getEditorShift().x;
    }

    default int translateToEditorY(int y) {
        return toEditorDimension(y) + getEditorShift().y;
    }

    default int translateFromEditorX(int x) {
        return fromEditorDimension(x - getEditorShift().x);
    }

    default int translateFromEditorY(int y) {
        return fromEditorDimension(y - getEditorShift().y);
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/SceneViewpoint.kt
package org.fbme.scenes.controllers

import java.awt.Point
import java.awt.Rectangle

interface SceneViewpoint {
    val editorShift: Point
    val editorScale: Float

    fun toEditorDimension(d: Int): Int {
        return (d * editorScale).toInt()
    }

    fun fromEditorDimension(d: Int): Int {
        return (d / editorScale).toInt()
    }

    fun translateToEditorX(x: Int): Int {
        return toEditorDimension(x) + editorShift.x
    }

    fun translateToEditorY(y: Int): Int {
        return toEditorDimension(y) + editorShift.y
    }

    fun translateFromEditorX(x: Int): Int {
        return fromEditorDimension(x - editorShift.x)
    }

    fun translateFromEditorY(y: Int): Int {
        return fromEditorDimension(y - editorShift.y)
    }

    fun fromEditor(bounds: Rectangle) = Rectangle(
        translateFromEditorX(bounds.x),
        translateFromEditorY(bounds.y),
        fromEditorDimension(bounds.width),
        fromEditorDimension(bounds.height)
    )
}
